hbase_sql/src/main/java/com/splicemachine/derby/stream/function/HFileGenerationFunction.java

```diff
@@ -268,4 +372,81 @@ private static InetSocketAddress getFavoredNode(BulkImportPartition partition) t
+            ExecRow execRow = operationContext.getOperation().getExecRowDefinition();
+            KeyEncoder keyEncoder = BulkImportUtils.getKeyEncoder(execRow, pkCols, tableVersion, null);
+            DataHash rowEncoder = BulkImportUtils.getRowHash(execRow, pkCols, tableVersion);
+            PairDecoder pairDecoder = new PairDecoder(keyEncoder.getDecoder(), rowEncoder.getDecoder(), execRow);
+            decoderMap.put(heapConglom, pairDecoder);
+        }
+        if (tentativeIndexList != null && tentativeIndexList.size() > 0) {
+            // For each index, initialize a decoder for index column
+            for (DDLMessage.TentativeIndex tentativeIndex : tentativeIndexList) {
+                DDLMessage.Table tbl = tentativeIndex.getTable();
+                DDLMessage.Index idx = tentativeIndex.getIndex();
+                Long indexConglomerate = idx.getConglomerate();
+                int n = idx.getIndexColsToMainColMapCount();
+                ExecRow execRow = new ValueRow(n);
+                List<Integer> formatIds = tbl.getFormatIdsList();
+                DataValueDescriptor[] dvds = execRow.getRowArray();
+                for (int i = 0; i < n; ++i) {
+                    dvds[i] =  LazyDataValueFactory.getLazyNull(formatIds.get(i));
+                }
+                int[] pkCols = new int[execRow.nColumns()];
+                for (int i = 0; i < pkCols.length; ++i) {
+                    pkCols[i] = i+1;
+                }
+                boolean[] sortOrder = new boolean[n];
+                for (int  i = 0; i < n; ++i) {
+                    sortOrder[i] = !idx.getDescColumns(i);
+                }
+                KeyEncoder keyEncoder = BulkImportUtils.getKeyEncoder(execRow, pkCols, tableVersion, sortOrder);
+                decoderMap.put(indexConglomerate, new Pair<>(execRow, keyEncoder.getDecoder()));
+            }
+        }
+    }
+
+    private void handleConstraintViolation(Long conglomerateId, byte[] key, byte[] value) throws StandardException{
+        ExecRow execRow;
+        if (isIndex(conglomerateId)) {
+            Pair<ExecRow, KeyDecoder> p = (Pair<ExecRow, KeyDecoder>)decoderMap.get(conglomerateId);
+            execRow = p.getFirst();
+            KeyDecoder decoder = (KeyDecoder) p.getSecond();
+            decoder.decode(key, 0, key.length, execRow);
+        }else {
+            PairDecoder decoder = (PairDecoder) decoderMap.get(conglomerateId);
+            execRow = decoder.decode(key, value);
+        }
+        if (operationContext != null && operationContext.isPermissive()) {
+
+            operationContext.recordBadRecord("Unique constraint violation, row: "
+                    + execRow.toString(), null);
+        }
+        else {
+            SpliceLogUtils.error(LOG, "Unique constraint violation, row: %s", execRow.toString());
+            DataDictionary dd = ah.getActivation().getLanguageConnectionContext().getDataDictionary();
```

Comment:
Isn't it possible to get the activation from the operationContext rather than passing around an ActivationHolder?

Reply:
Not always possible. If it is not a SpliceOperation (in case create index), Activation will not be serialized

Reply:
Is it possible to create an operationContext with a dummy operation and make "create index" use it so that we always have an operationContext?

Reply:
It is possible, but I think it's a bit complicated to fake a SpliceOperation than serde activation.